proxy.go

```diff
@@ -232,19 +190,39 @@ func executeWithRetry(
 		}
 		// The request has been successfully proxied.
 
-		since = time.Since(startTime).Seconds()
-		monitorDuration(since)
-
 		srw.SetStatusCode(rw.StatusCode())
 		// StatusBadGateway response is returned by http.ReverseProxy when
 		// it cannot establish connection to remote host.
 		if rw.StatusCode() == http.StatusBadGateway {
 			log.Debugf("the invalid host is: %s", s.host.addr)
 			s.host.penalize()
-			q := getQuerySnippet(req)
-			err1 := fmt.Errorf("%s: cannot reach %s; query: %q", s, s.host.addr.Host, q)
-			respondWith(srw, err1, srw.StatusCode())
+			s.host.dec()
+			if numRetry < maxRetry && s.host.replica.cluster.getHost().isActive() {
+				shouldRetry = true
+				// the query execution has been failed
+				atomic.StoreUint32(&s.host.active, uint32(0))
+				monitorRetryRequestInc(s.labels)
+
+				// update host
+				h := s.host
+				s.host = h.replica.cluster.getHost()
+
+				req.URL.Host = s.host.addr.Host
+				req.URL.Scheme = s.host.addr.Scheme
+				log.Debugf("the valid host is: %s", s.host.addr)
+			} else {
+				shouldRetry = false
+				since = time.Since(startTime).Seconds()
+				monitorDuration(since)
+				q := getQuerySnippet(req)
+				err1 := fmt.Errorf("%s: cannot reach %s; query: %q", s, s.host.addr.Host, q)
+				respondWith(srw, err1, srw.StatusCode())
+			}
+		} else {
+			since = time.Since(startTime).Seconds()
+			return since, nil
```

Comment:
it feels weird that inside the loop you end it with `return since, nil` while a few line before (L214) you end the loop with a `shouldRetry = false` . I'm not saying one way is better than the other one but your code is inconsistent since you're using 2 ways for the same goal just a few lines apart

Reply:
Perhaps @JingmaoYou it'll be easier if you start with an never ending loop:
```
for {...}
```
and control the exit with return statements